aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -589,6 +589,132 @@ Tensor gather_backward(const Tensor& grad, const Tensor& self, int64_t dim, cons
+  // This replicates behavior in (https://github.com/pytorch/pytorch/pull/13781)
+  // needed to allow workflows with empty batch to succeed.
+  if (data_dims[axis] == 0) {
+    shape.insert(shape.end(), data_dims.begin(), data_dims.end());
+  } else {
+    shape.insert(shape.end(), data_dims.begin(), data_dims.begin() + axis);
+    shape.insert(shape.end(), indices_dims.begin(), indices_dims.end());
+    shape.insert(shape.end(), data_dims.begin() + axis + 1, data_dims.end());
+  }
+  return shape;
+}
+// Check that indices fall within dimension array size
+template <typename IndexType>
+static void check_indexarray_range(
+    const IndexType* indices,
+    int64_t n,
+    IndexType indexing_axis_dim) {
+  //
+  for (auto i = 0; i < n; ++i) {
+    auto idx = indices[i];
+    if (idx < 0) {
+      idx = idx + indexing_axis_dim;
+    }
+    TORCH_CHECK(
+        0 <= idx && idx < indexing_axis_dim,
+        "INDICES element is out of DATA bounds, id=",
+        idx,
+        " axis_dim=",
+        indexing_axis_dim);
+  }
+}
+Tensor & batch_index_select_out_cpu(Tensor & result, const Tensor & self, const Tensor & index) {
+  result.resize_({self.sizes()[0], index.sizes()[0]});
+
+  auto self_contig = self.contiguous();
+  auto index_contig = index.contiguous();
+  const caffe2::TypeMeta dataType = self_contig.dtype();
+  size_t item_bytesize = dataType.itemsize();
+
+  TORCH_CHECK(self_contig.dim() >= 2, "DATA should be at least 2D");
+
+  // New shape:
+  //  [data dims before axis] + [indices dims] + [data dims after axis]
+  std::vector<int64_t> shape = calc_output_shape_vector<int64_t>(
+      self_contig.sizes(), index_contig.sizes(), 1);
+  Tensor* output = &result;
+  auto out = static_cast<char*>(output->data_ptr());
+
+  // Succeed if size of output is zero, which can happen for empty batch which
+  // would have data dimension size of 0.
+  if (output->numel() == 0) {
+    return result;
+  }
+
+  auto src_base = static_cast<const char*>(self_contig.data_ptr());
+
+  auto outer_dims_product = c10::size_to_dim_(1, self_contig.sizes());
+  auto block_size = c10::size_from_dim_(1 + 1, self_contig.sizes());
+  auto block_bytesize = block_size * item_bytesize;
+
+  auto src_indexing_axis_dim = self_contig.size(1);
+  auto src_batch_bytesize = c10::size_from_dim_(1, self_contig.sizes()) * item_bytesize;
+  // Treat indices as a single block even if they have multiple dimensions.
+  // The "gathered batch" is a cumulative result combining indexed blocks.
+  auto idx_inner_dims_product = c10::size_from_dim_(1, index_contig.sizes());
+  auto N = index_contig.numel();
+
+  auto gathered_batch_bytesize = N * block_size * item_bytesize;
+
+  AT_DISPATCH_INDEX_TYPES(
+    index_contig.scalar_type(), "batch_gather_compute", [&]() {
+
+      const auto* idxs = index_contig.data_ptr<index_t>();
+      check_indexarray_range<index_t>(idxs, N, src_indexing_axis_dim);
+
+      // Special-case single-float copy for efficiency
```

Comment:
pt tensors have a fixed set of types, so you probably just want to dispatch based on that (and thus have faster passes for all of them)